src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -2699,6 +2781,10 @@ public void close() throws SQLServerException {
             tdsChannel.close();
         }
 
+        // Invalidate statement cache.
+        if(null != this.preparedStatementCache)
+            this.preparedStatementCache.clear();
```

Comment:
The `EvictionListener` is not called on an explicit removal, only automatic (hence the switch to `RemovalListener` in Guava). You probably want to iterate the keys, remove, and forward the value to your discard logic.